File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include <vector>
#include <queue>

using namespace std;

int N,M;
int dir[4][2]={{0,1},{1,0},{0,-1},{-1,0}};
int arr[104][104];
int visited[104][104];

void BFS_iter(int x,int y) {
    int dx,dy;
    queue<pair<int,int>> q;
    q.push({x,y});
    visited[x][y]=1;
    
    while(!q.empty()) {
        x=q.front().first;
        y=q.front().second;
        q.pop();
        
        for(int i=0;i<4;i++) {
            dx=x+dir[i][0];
            dy=y+dir[i][1];
            
            if(dx<0 or dx>=N or dy<0 or dy>=M or arr[dx][dy]==0) {
                continue;
            }
            
            if(visited[dx][dy]==0) {
                visited[dx][dy]=visited[x][y]+1;
                q.push({dx,dy});
            }
        }
    }
}
int solution(vector<vector<int> > maps)
{
    int answer = 0;
    N=maps.size();
    M=maps[0].size();
    
    for(int i=0;i<N;i++) {
        for(int j=0;j<M;j++) {
            arr[i][j]=maps[i][j];
        }
    }
    
    BFS_iter(0,0);
    
    if(visited[N-1][M-1]) {
        return visited[N-1][M-1];
    }
    else {
        return -1;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    
    static int N,M;
    static int[][]arr=new int[104][104];
    static int[][]visited=new int[104][104];
    static int[][] dir={{0,1},{1,0},{0,-1},{-1,0}};
    static Queue<int[]> q=new LinkedList<>();
    
    public void BFS_iter(int x,int y) {
        
        int dx,dy;
        q.add(new int[]{x,y});
        visited[x][y]=1;
        
        while(!q.isEmpty()) {
            int []temp=q.poll();
            x=temp[0];
            y=temp[1];
            
            for(int i=0;i<4;i++) {
                dx=x+dir[i][0];
                dy=y+dir[i][1];
                
                if(dx<0 || dx>=N || dy<0 || dy>=M || arr[dx][dy]==0) {
                    continue;
                }
                
                if(visited[dx][dy]==0) {
                    visited[dx][dy]=visited[x][y]+1;
                    q.add(new int[]{dx,dy});
                }
            }
        }
    }
    
    public int solution(int[][] maps) {
        int answer = 0;
        arr=maps;
        N=maps.length;
        M=maps[0].length;
        
        BFS_iter(0,0);
        answer=visited[N-1][M-1];
        
        
        return answer==0?-1:answer;
    }
}